opentelemetry-api/opentelemetry/backend.py

```diff
@@ -0,0 +1,131 @@
+       import a module with that name and call a function `get_opentelemetry_backend_impl` in it.
+       The function receives the API type that is expected as an argument and should return an
+       instance of it (e.g., the argument is :cls:`opentelemetry.trace.Tracer` and the function should
+       return an instance of a :cls:`~Tracer` (probably of a derived type).
+    2. If the variable is not set, `$OPENTELEMETRY_PYTHON_BACKEND_DEFAULT` is tried instead.
+    3. If that variable was also not set, an attempt is made to import and use the OpenTelemetry
+       SDK.
+    4. Otherwise (if no variable was set and the SDK was not importable, or an error occured when
+       trying to instantiate the implementation object) the default implementation that ships with the
+        API distribution (a fast no-op implementation) is used.
+
+If you called the setter or if you call the setter later (not recommended), the object you set is
+used instead.
+"""
+
+import sys
+import os
+import importlib
+from typing import TypeVar, Type, Callable
+from functools import wraps
+
+from .trace import Tracer
+
+### Generic private code {{{1
+
+_T = TypeVar('_T')
+
+_DEFAULT_BACKEND_MODNAME = 'opentelemetry.sdk.internal.backend_impl'
+_UNIT_TEST_IGNORE_ENV = False
+
+_tracer = None
+
+def _get_fallback_impl(api_type: Type[_T]) -> _T:
+    # TODO: Move (most of) this to module docstring.
+    """Gets the fallback implementation for ``api_type``.
+
+    ``api_type`` must be a OpenTelemetry API type like :cls:`Tracer`.
+
+    First, the function tries to find a module that provides a `get_opentelemetry_backend_impl`
+    function (with the same signature as this function). The following modules are tried:
+
+    1. `$OPENTELEMETRY_PYTHON_BACKEND_<typename>` (e.g. `OPENTELEMETRY_PYTHON_BACKEND_TRACER`)
+    2. `$OPENTELEMETRY_PYTHON_BACKEND_DEFAULT` (e.g. `OPENTELEMETRY_PYTHON_BACKEND_TRACER`)
+    3. The OpenTelemetry SDK's tracer module.
+
+    Note that if any of the environment variables is set to an nonempty value, further steps
+    are not tried, even if the modulename set there is invalid or fails to load. The no-op API
+    implementation is returned instead.
+    """
+    backend_modname = None
+    if not _UNIT_TEST_IGNORE_ENV and not sys.flags.ignore_environment:
+        backend_modname = os.getenv('OPENTELEMETRY_PYTHON_BACKEND_' + api_type.__name__.upper())
+        if not backend_modname:
+            backend_modname = os.getenv('OPENTELEMETRY_PYTHON_BACKEND_DEFAULT')
+    if not backend_modname:
+        backend_modname = _DEFAULT_BACKEND_MODNAME
+    try:
+        backend_mod = importlib.import_module(backend_modname)
+    except (ImportError, SyntaxError):
+        # TODO Log/warn
+        return api_type()
+    try:
+        # Note: We use such a long name to avoid calling a function that is not intended for this
+        # API.
+        backend_fn: Callable[[Type[_T]], object] = getattr(backend_mod, 'get_opentelemetry_backend_impl')
```

Comment:
This is failing in 3.4. In general I don't think we need type annotations outside API methods.

Reply:
It was failing with `mypy --strict` without that. But if it passes with your `mypy.ini` from #31, that's fine.